vision/segmentation/plane_segmentation.go

```diff
@@ -66,6 +66,61 @@ func pointCloudSplit(cloud pc.PointCloud, inMap map[r3.Vector]bool) (pc.PointClo
 	return mapCloud, nonMapCloud, nil
 }
 
+// SegmentPlaneWRTGround segments the biggest 'ground' plane in the 3D Pointcloud.
+// nIterations is the number of iteration for ransac
+// nIter to choose? nIter = log(1-p)/log(1-(1-e)^s), where p is prob of success, e is outlier ratio, s is subset size (3 for plane).
+// dstThreshold is the float64 value for the maximum allowed distance to the found plane for a point to belong to it
+// This function returns a Plane struct, as well as the remaining points in a pointcloud
+// It also returns the equation of the found plane: [0]x + [1]y + [2]z + [3] = 0.
+// angleThrehold is the maximum acceptable angle between the groundVec and angle of the plane,
+// if the plane is at a larger angle than maxAngle, it will not be considered for segmentation, if set to 0 then not considered
+// normalVec is the normal vector of the plane representing the ground.
+func SegmentPlaneWRTGround(ctx context.Context, cloud pc.PointCloud, nIterations int, angleThreshold,
+	dstThreshold float64, normalVec r3.Vector,
+) (pc.Plane, pc.PointCloud, error) {
+	if cloud.Size() <= 3 { // if point cloud does not have even 3 points, return original cloud with no planes
+		return pc.NewEmptyPlane(), cloud, nil
+	}
+	//nolint:gosec
+	r := rand.New(rand.NewSource(1))
+	pts := GetPointCloudPositions(cloud)
+	nPoints := cloud.Size()
+
+	// First get all equations
+	equations := make([][4]float64, 0, nIterations)
+	for i := 0; i < nIterations; i++ {
+		// sample 3 Points from the slice of 3D Points
+		n1, n2, n3 := utils.SampleRandomIntRange(1, nPoints-1, r),
+			utils.SampleRandomIntRange(1, nPoints-1, r),
+			utils.SampleRandomIntRange(1, nPoints-1, r)
+		p1, p2, p3 := pts[n1], pts[n2], pts[n3]
+
+		// get 2 vectors that are going to define the plane
+		v1 := p2.Sub(p1)
+		v2 := p3.Sub(p1)
+		// cross product to get the normal unit vector to the plane (v1, v2)
+		cross := v1.Cross(v2)
+		planeVec := cross.Normalize()
+		// find current plane equation denoted as:
+		// cross[0]*x + cross[1]*y + cross[2]*z + d = 0
+		// to find d, we just need to pick a point and deduce d from the plane equation (vec orth to p1, p2, p3)
+		d := -planeVec.Dot(p2)
+
+		currentEquation := [4]float64{planeVec.X, planeVec.Y, planeVec.Z, d}
+
+		if angleThreshold != 0 {
+			if math.Acos(normalVec.Dot(planeVec)) <= angleThreshold*math.Pi/180.0 {
+				equations = append(equations, currentEquation)
+			} else {
+				i--
+			}
```

Comment:
I don't think you need this. I suspect it may extend the time you are searching by quite a bit. Even though many of the planes you match will not fall within the ground plane limits, it's not necessary to always have N many "viable" candidates. May of them will end up being repeats.

Reply:
Agreed, but since the array of equations is already made, we will probably be looping over all entries in the findBestEq function. I assume a lot of the array would then be blank, so I'm going to skip the equation 0x + 0y + 0z = 0.